File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    private void dfs(int N,int city,int isConnected[][],boolean visited[]){
        visited[city]=true;
        for (int othCity=0;othCity<N;othCity++){
            if(!visited[othCity] && isConnected[city][othCity]==1){
                dfs(N,othCity,isConnected,visited);
            }
        }
    }
    public int findCircleNum(int[][] isConnected) {
        int N=isConnected.length,province=0;
        boolean visited[]=new boolean[N];
        for(int city=0;city<N;city++){
            if(!visited[city]){
                province++;
                dfs(N,city,isConnected,visited);
            }
        }
        return province;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def dfs(self, N, city, isConnected, visited):
        visited[city] = True  # Mark city as visited
        for othCity in range(N):
            if not visited[othCity] and isConnected[city][othCity] == 1:
                self.dfs(N, othCity, isConnected, visited)

    def findCircleNum(self, isConnected: list[list[int]]) -> int:
        N = len(isConnected)
        province = 0
        visited = [False] * N  # Boolean array to track visited cities

        for city in range(N):
            if not visited[city]:  # If the city is not visited, start DFS
                province += 1
                self.dfs(N, city, isConnected, visited)

        return province
